src/ray/raylet/runtime_env_agent_client.cc

```diff
@@ -0,0 +1,395 @@
+    try_invoke_once(succ_callback, [=](ray::Status status) {
+      if (status.IsNotFound()) {
+        RAY_LOG(INFO) << "Scheduling a retry in "
+                      << RayConfig::instance().agent_manager_retry_interval_ms()
+                      << "ms...";
+        this->delay_executor_(
+            [=]() { RetryInvoke(try_invoke_once, succ_callback, fail_callback); },
+            RayConfig::instance().agent_manager_retry_interval_ms());
+      } else {
+        // Non retryable errors, invoke fail_callback
+        fail_callback(std::move(status));
+      }
+    });
+  }
+
+  // Making HTTP call.
+  // POST /get_or_create_runtime_env
+  // Body = proto rpc::GetOrCreateRuntimeEnvRequest
+  void GetOrCreateRuntimeEnv(const JobID &job_id,
+                             const std::string &serialized_runtime_env,
+                             const rpc::RuntimeEnvConfig &runtime_env_config,
+                             const std::string &serialized_allocated_resource_instances,
+                             GetOrCreateRuntimeEnvCallback callback) override {
+    RetryInvoke<rpc::GetOrCreateRuntimeEnvReply>(
+        [=](SuccCallback<rpc::GetOrCreateRuntimeEnvReply> succ_callback,
+            FailCallback fail_callback) {
+          return TryGetOrCreateRuntimeEnv(job_id,
+                                          serialized_runtime_env,
+                                          runtime_env_config,
+                                          serialized_allocated_resource_instances,
+                                          succ_callback,
+                                          fail_callback);
+        },
+        /*succ_callback=*/
+        [=](rpc::GetOrCreateRuntimeEnvReply reply) {
+          if (reply.status() != rpc::AGENT_RPC_STATUS_OK) {
+            RAY_LOG(INFO) << "Failed to create runtime env for job " << job_id
+                          << ", error message: " << reply.error_message();
+            RAY_LOG(DEBUG) << "Serialized runtime env for job " << job_id << ": "
+                           << serialized_runtime_env;
+            callback(false,
+                     reply.serialized_runtime_env_context(),
+                     /*setup_error_message*/ reply.error_message());
+          } else {
+            RAY_LOG(INFO) << "Create runtime env for job " << job_id;
+            callback(true,
+                     reply.serialized_runtime_env_context(),
+                     /*setup_error_message*/ "");
+          }
+        },
+        /*fail_callback=*/
+        [=](ray::Status status) {
+          RAY_LOG(INFO)
+              << "Failed to create runtime env for job " << job_id
+              << ", status = " << status
+              << ", maybe there are some network problems, will fail the request.";
+          RAY_LOG(DEBUG) << "Serialized runtime env for job " << job_id << ": "
+                         << serialized_runtime_env;
+          callback(false, "", "Failed to request agent.");
```

Comment:
I believe this will bubble up to the user level exception upon runtime env failure? So we need a better error message. Can you write the status code + detailed error here instead? (e.g., maybe there are some network problems or the runtime env agent failed to start. status.code(), status.message())